improver/memprofile.py

```diff
@@ -0,0 +1,125 @@
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing maximum memory profiling utilities."""
+
+import tracemalloc as tm
+from queue import Queue, Empty
+from threading import Thread
+from resource import getrusage, RUSAGE_SELF
+from datetime import datetime
+
+
+def memory_profile_start(outfile_prefix, timeout_time=0.1):
+    """Starts the memory tracking profiler.
+
+    Args:
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+        timeout_time (float):
+            Time the queue will look for an input.
+
+    Returns:
+        Active Thread tracking the memory.
+        Active Queue for communication to the thread.
+    """
+    queue = Queue()
+    thread = Thread(target=memory_monitor, args=(queue, timeout_time,
+                                                 outfile_prefix))
+    thread.start()
+    return thread, queue
+
+
+def memory_profile_end(queue, thread):
+    """Ends the memory tracking profiler.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with active thread.
+        thread (thread.Thread):
+            Active thread instance running memory tracking.
+    """
+    queue.put('Stop')
+    thread.join()
+
+
+def memory_monitor(queue, timeout_time, outfile_prefix):
+    """Function to track memory usage, should be run in a separate
+    thread to the main program.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with the thread.
+        timeout_time (float):
+            Time the queue will look for in input.
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+    """
+    tm.start()
+    old_max = 0
+    snapshot = None
+
+    fout = open("{}_MAX_TRACKER".format(outfile_prefix), 'w')
+    b2mb = 1 / 1048576
+
+    while True:
+        try:
+            queue.get(timeout=timeout_time)
+            snapshot.dump("{}_SNAPSHOT".format(outfile_prefix))
```

Comment:
it might be worth seeing how long the snapshot dump takes on a larger plugin (~GB) to give a better idea of what the default sample frequency should be.

Reply:
The dump only happens once the main thread has finished, so would it matter how long it takes?

Reply:
my mistake, I was misreading this